File: src/ipc/events.ts
import { type UnlistenFn, listen } from "@tauri-apps/api/event";

// Tauri event name constants. Backend counterparts are in src-tauri/src/ipc/events.rs.
export const CLI_LOST = "cli:lost" as const;

/** Event emitted by the backend to display a toast notification in the frontend.
 *  Payload shape: `{ kind: "error"|"warning"|"info"|"success", title: string, body: string }`.
 *  See KB §5.5 and src-tauri/src/ipc/events.rs TOAST_SHOW. */
export const TOAST_SHOW = "toast:show" as const;

/** Emitted after each git poll cycle for a single project.
 *  Payload shape: `{ id: string, status: GitStatus }`.
 *  Backend counterpart: `GIT_UPDATED` in src-tauri/src/ipc/events.rs. */
export const GIT_UPDATED = "git:updated" as const;

/** Emitted once when a run transitions from Pending to Running.
 *  Payload shape: `{ run_id: string, project_id: string }`.
 *  Backend counterpart: `RUN_STARTED` in src-tauri/src/ipc/events.rs. */
export const RUN_STARTED = "run:started" as const;

/** Emitted for each structured event parsed from the child's stdout.
 *  Payload shape: `{ run_id: string, event: RunEvent }`.
 *  Backend counterpart: `RUN_EVENT` in src-tauri/src/ipc/events.rs. */
export const RUN_EVENT = "run:event" as const;

/** Emitted once when a run reaches a terminal state (Completed, Failed, or Stopped).
 *  Payload shape: `{ run_id: string, status: RunStatus, exit_code: number | null }`.
 *  Backend counterpart: `RUN_FINISHED` in src-tauri/src/ipc/events.rs. */
export const RUN_FINISHED = "run:finished" as const;

export function subscribe<T>(
  eventName: string,
  handler: (payload: T) => void
): Promise<UnlistenFn> {
  return listen<T>(eventName, (event) => handler(event.payload));
}

File: src-tauri/src/ipc/events.rs
/// Tauri event name constants emitted by the backend.
/// Frontend counterparts are in `src/ipc/events.ts`.
pub const CLI_LOST: &str = "cli:lost";

/// Emitted after each git poll cycle for a single project.
///
/// Payload shape: `{ id: string, status: GitStatus }`.
/// Frontend counterpart: `GIT_UPDATED` in `src/ipc/events.ts`.
pub const GIT_UPDATED: &str = "git:updated";

/// Event emitted to display a toast notification in the frontend.
///
/// Payload shape (KB §5.5): `{ kind, title, body, run_id? }`
/// where `kind` is one of `"error"`, `"warning"`, `"info"`, `"success"`.
pub const TOAST_SHOW: &str = "toast:show";

/// Emitted once when a run transitions from `Pending` to `Running`.
///
/// Payload shape: `{ run_id: string, project_id: string }`.
/// Frontend counterpart: `RUN_STARTED` in `src/ipc/events.ts`.
pub const RUN_STARTED: &str = "run:started";

/// Emitted for each structured event parsed from the child's stdout.
///
/// Payload shape: `{ run_id: string, event: RunEvent }`.
/// Frontend counterpart: `RUN_EVENT` in `src/ipc/events.ts`.
pub const RUN_EVENT: &str = "run:event";

/// Emitted once when a run reaches a terminal state (Completed, Failed, or Stopped).
///
/// Payload shape: `{ run_id: string, status: RunStatus, exit_code: number | null }`.
/// Frontend counterpart: `RUN_FINISHED` in `src/ipc/events.ts`.
pub const RUN_FINISHED: &str = "run:finished";

/// Emitted when the parser detects a step-failure sentinel in the child's stdout.
///
/// Payload shape: `{ run_id: string }`.
/// The UI should surface a Retry / Skip / Abort / Continue prompt on this event.
/// A 60 s auto-Continue fires if no `respond_to_step_failure` command arrives.
pub const RUN_STEP_FAILURE: &str = "run:step_failure";
